### PR TITLE
Implement pre-render scanline in ppu

### PR DESCRIPTION
**Pull request by johnor**
First steps. ~~Some TODOs that will be fixed in follow up PRs. Will add more tests later also.~~

### CODE REVIEW

File: core/src/ppu.cpp [hunk 1]
```diff
@@ -222,4 +295,116 @@ void Ppu::clear_vblank_flag() {
...
+            registers_->attribute_table_shifter_hi =
+                    (registers_->attribute_table_shifter_hi & 0xFF00u) |
+                    ((registers_->attribute_table_latch & 0b10u) ? 0xFFu
+                                                                 : 0x00u);
+
+            registers_->name_table = registers_->name_table_latch;
+            const uint16_t nametable_address =
+                    get_nametable_address(registers_->vram_addr);
+            registers_->name_table_latch = mmu_->read_byte(nametable_address);
+        } break;
```
Comment (robinlinden) [inline core/src/ppu.cpp:367]: The formatting of the breaks in this switch are pretty odd? Is this intentional? :O I'm not hating it, but I'm surprised clang-format doesn't care.

**Review by robinlinden — APPROVED** [review]
I think everything looks good, but it's some pretty dense code, so it might be good if @mkiael or @sa12123 also has a look.

File: core/src/ppu.cpp [hunk 2]
```diff
@@ -22,6 +22,76 @@ const uint16_t kPostRenderScanline = 240;
 const uint16_t kVBlankScanlineStart = 241;
 const uint16_t kVBlankScanlineEnd = 260;
 
+uint16_t increase_vram_coarse_x(uint16_t vram_addr) {
```
[the remark this answers is not in the document]
Comment (johnor) [reply core/src/ppu.cpp]: Done!

File: core/src/ppu.cpp (hunk 1 above)
Comment (johnor) [reply core/src/ppu.cpp]: I put moved break; to inside the brackets now. Is that better?

Comment (robinlinden) [reply core/src/ppu.cpp]: I didn't mind it before either. I was mostly surprised clang-format didn't care. :P

**Review by robinlinden — APPROVED** [review]

**Pull request merged**